spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java

```diff
@@ -211,9 +233,21 @@ public RecordMetadata syncSend(String topic, int partition, K key, V data)
 		return future;
 	}
 
-	@Override
-	public void flush() {
-		this.producer.flush();
+	private ProducerRecord<K, V> messageToProducerRecord(Message<?> message) {
+		MessageHeaders headers = message.getHeaders();
+		Object topic = headers.get(KafkaHeaders.TOPIC);
+		Object partition = headers.get(KafkaHeaders.PARTITION_ID);
+		Object key = headers.get(KafkaHeaders.MESSAGE_KEY);
+		Object payload = message.getPayload();
+		Assert.isTrue(topic == null || topic instanceof String, "topic header must be a String");
+		Assert.isTrue(partition == null || partition instanceof Integer, "partition header must be an Integer");
```

Comment:
Can't we use an overloaded version of `MessageHeaders.get()` with the `Class` to cast instead?